subsys/bluetooth/host/classic/hfp_hf.c

```diff
@@ -391,732 +393,2065 @@ int cind_resp(struct at_client *hf_at, struct net_buf *buf)
+{
+	struct bt_hfp_hf *hf = CONTAINER_OF(hf_at, struct bt_hfp_hf, at);
+	uint32_t inband;
+	int err;
+
+	err = at_get_number(hf_at, &inband);
+	if (err) {
+		LOG_ERR("could not get bsir value");
+		return err;
+	}
+
+	if (inband > 1) {
+		LOG_ERR("Invalid %d bsir value", inband);
+		return -EINVAL;
+	}
+
+	if (bt_hf->inband_ring) {
+		bt_hf->inband_ring(hf, (bool)inband);
+	}
+
+	return 0;
+}
+
 #if defined(CONFIG_BT_HFP_HF_CODEC_NEG)
-	{send_at_bac, at_bac_finish, false},
-#endif /* CONFIG_BT_HFP_HF_CODEC_NEG */
-};
+int bcs_handle(struct at_client *hf_at)
+{
+	struct bt_hfp_hf *hf = CONTAINER_OF(hf_at, struct bt_hfp_hf, at);
+	uint32_t codec_id;
+	int err;
+
+	err = at_get_number(hf_at, &codec_id);
+	if (err) {
+		LOG_ERR("could not get bcs value");
+		return err;
+	}
+
+	if (!(hf->hf_codec_ids & BIT(codec_id))) {
+		LOG_ERR("Invalid codec id %d", codec_id);
+		err = bt_hfp_hf_set_codecs(hf, hf->hf_codec_ids);
+		return err;
+	}
+
+	atomic_set_bit(hf->flags, BT_HFP_HF_FLAG_CODEC_CONN);
+
+	if (bt_hf->codec_negotiate) {
+		bt_hf->codec_negotiate(hf, codec_id);
+		return 0;
+	}
+
+	err = bt_hfp_hf_select_codec(hf, codec_id);
+	return err;
+}
+#endif /* CONFIG_BT_HFP_HF_CODEC_NEG */
+
+static int btrh_handle(struct at_client *hf_at)
+{
+	struct bt_hfp_hf *hf = CONTAINER_OF(hf_at, struct bt_hfp_hf, at);
+	uint32_t on_hold;
+	int err;
+	struct bt_hfp_hf_call *call;
+
+	err = at_get_number(hf_at, &on_hold);
+	if (err < 0) {
+		LOG_ERR("Error getting value");
+		return err;
+	}
+
+	if (on_hold == BT_HFP_BTRH_ON_HOLD) {
+		call = get_call_with_state(hf, BT_HFP_HF_CALL_STATE_WAITING);
+		if (!call) {
+			return 0;
+		}
+
+		if (!atomic_test_bit(call->flags, BT_HFP_HF_CALL_INCOMING)) {
+			LOG_WRN("Invalid call dir (outgoing)");
+		}
+		atomic_set_bit(call->flags, BT_HFP_HF_CALL_INCOMING_HELD);
+	} else {
+		call = get_call_with_state_and_flag(hf,
+			BT_HFP_HF_CALL_STATE_ACTIVE, BT_HFP_HF_CALL_INCOMING_HELD);
+		if (!call) {
+			return 0;
+		}
+
+		if (on_hold == BT_HFP_BTRH_ACCEPTED) {
+			atomic_clear_bit(call->flags, BT_HFP_HF_CALL_INCOMING_HELD);
+			if (bt_hf && bt_hf->accept) {
+				bt_hf->accept(call);
+			}
+		} else if (on_hold == BT_HFP_BTRH_REJECTED) {
+			hf_reject_call(call);
+		} else {
+			return -EINVAL;
+		}
+	}
+
+	return 0;
+}
+
+#if defined(CONFIG_BT_HFP_HF_3WAY_CALL)
+static int ccwa_handle(struct at_client *hf_at)
+{
+	struct bt_hfp_hf *hf = CONTAINER_OF(hf_at, struct bt_hfp_hf, at);
+	char *number;
+	uint32_t type;
+	int err;
+	struct bt_hfp_hf_call *call;
+
+	number = at_get_string(hf_at);
+
+	err = at_get_number(hf_at, &type);
+	if (err) {
+		LOG_WRN("could not get the type");
+	} else {
+		type = 0;
+	}
+
+	call = get_new_call(hf);
+	if (!call) {
+		LOG_ERR("Not available call object");
+		return 0;
+	}
+
+	hf_call_state_update(call, BT_HFP_HF_CALL_STATE_INCOMING);
+	if (bt_hf->call_waiting) {
+		bt_hf->call_waiting(call, number, (uint8_t)type);
+	}
+
+	return 0;
+}
+
+static struct _chld_feature {
+	char *name;
```

Comment:
Seems like `const char *name` is more appropriate, since all the values in the subsequent array are string literals.

Reply:
updated